### ISSUE TITLE
How to implement a python-based motion planning algorithm in moveit?

### ISSUE THREAD

**Issue by Evansika959** [ros-planning/moveit#3273]
Moveit has many motion planning library, but to my knowledge they're all implemented by cpp. We wonder if there is any python-based motion-planning library(like OMPL in py version) so that we could implement our python-based motion planning algorithm in moveit? 

Thank you in advance!!

**Comment by rhaschke** [comment]
To my knowledge, there is no way to implement a scheduling plugin in Python (yet). Since the planning is usually very time-consuming, this is also not recommended.

**Comment by Evansika959** [comment]
> To my knowledge, there is no way to implement a scheduling plugin in Python (yet). Since the planning is usually very time-consuming, this is also not recommended.

thanks! cpp is indeed more suitble for planning

**Comment by gavanderhoorn** [comment]
While I agree with @rhaschke about the potential problems with performance, I've always been curious whether [embedding a Python interpreter](https://docs.python.org/3/extending/embedding.html) (perhaps using something like [reefactor/cppy3](https://github.com/reefactor/cppy3)) could not be used host an interpreter in a planning plugin.

Then it'd be a matter of exposing the right amount of state (from MoveIt to the Python3 interpreter) and writing the actual business logic (ie: planning algorithm).

**Comment by rhaschke** [comment]
You are right, using pybind11 as in #2910 and exposing some more state, this could become possible.

**Comment by gavanderhoorn** [comment]
Ah, `pybind11` can also embed. I forgot.

**Comment by Evansika959** [comment]
Thank you for sharing!!

**Comment by Evansika959** [comment]
I wonder if there're actual projects done by implementing python planning program in this manner? 
Or some simple examples of embedding py planning programs in moveit could also be a great help!
Thanks! 

> While I agree with @rhaschke about the potential problems with performance, I've always been curious whether [embedding a Python interpreter](https://docs.python.org/3/extending/embedding.html) (perhaps using something like [reefactor/cppy3](https://github.com/reefactor/cppy3), or even the well-known [boost.python](https://www.boost.org/doc/libs/1_80_0/libs/python/doc/html/index.html)) could not be used host an interpreter in a planning plugin.
> 
> Then it'd be a matter of exposing the right amount of state (from MoveIt to the Python3 interpreter) and writing the actual business logic (ie: planning algorithm).

**Comment by rhaschke** [comment]
I'm not aware of any python-based planner for MoveIt.
